src/certgen_certificate_generator_server.cc

```diff
@@ -27,11 +27,285 @@
 */
 
 #include "fty_certificate_generator_classes.h"
+#include <chrono>
+#include <netdb.h>
+
+// because there is NO std::chrono::days in C++14
+#define HOURS_IN_DAY        24
+
+#define COMMON_NAME         "Web server-ipm2-jana-selfsigned"
+#define ORGANIZATION_UNIT   "Power quality"
+#define ORGANIZATION        "Eaton"
+#define LOCALITY            "Grenoble"
+#define STATE               "38"
+#define COUNTRY             "FR"
+#define EMAIL               "noemail"
 
 //  Structure of our class
 
 namespace certgen
 {
+    using namespace fty;
+    using namespace std::placeholders;
+
+    CertificateGeneratorServer::CertificateGeneratorServer()
+    {
+        //initiate the commands handlers
+        m_supportedCommands[GENERATE_SELFSIGNED_CERTIFICATE] = std::bind(&CertificateGeneratorServer::handleGenerateSelfsignedCertificate, this, _1);
+        m_supportedCommands[GENERATE_CSR] = std::bind(&CertificateGeneratorServer::handleGenerateCSR, this, _1);
+        m_supportedCommands[IMPORT_CERTIFICATE] = std::bind(&CertificateGeneratorServer::handleImportCertificate, this, _1);
+    }
+
+    Payload CertificateGeneratorServer::handleRequest(const Sender & /*sender*/, const Payload & payload)
+    {
+        try
+        {
+            if (payload.size() == 0)
+            {
+                throw std::runtime_error("Command frame is empty");
+            }
+
+            Command cmd = payload.at(0);
+            if (cmd == "ERROR" || cmd == "OK")
+            {
+                //avoid loop
+                return {};
+            }
+
+            //check if the command exists in the system
+            if (m_supportedCommands.count(cmd) == 0)
+            {
+                throw std::runtime_error(cmd + "not supported");
+            }
+            FctCommandHandler cmdHandler = m_supportedCommands[cmd];
+
+            //create copy of the payload
+            std::vector<std::string> params(payload.begin() + 1, payload.end());
+
+            std::string result = cmdHandler(params);
+            return {result};
+        }
+        catch (std::exception &e)
+        {
+            log_error("Unexpected error: %s", e.what());
+            return {"ERROR", e.what()};
+        }
+        catch (...) //show must go one => Log and ignore the unknown error
+        {
+            log_error("Unexpected error: unknown");
+            return {"ERROR", ""};
+        }
+    }
+
+    std::list<std::string> CertificateGeneratorServer::getSystemDomainNames()
```

Comment:
see https://www.gnu.org/software/libc/manual/html_node/Host-Identification.html to get the fqdn

Reply:
The described approach reads domain name from /etc/hosts first, where is the hostname withoout domain - so not the one we need.
I used already existing environment variable in the end.